File: ros2_ws/src/param_demo_cpp/src/param_demo.cpp
#include <algorithm>
#include <chrono>
#include <memory>
#include <string>
#include <thread>
#include <vector>

#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

class ParamDemoNode : public rclcpp::Node
{
public:
  ParamDemoNode()
  : Node("param_demo_cpp"), done_(false), loop_count_(0)
  {
    declare_parameter<int>("param1", 111);
    declare_parameter<int>("param2", 222);
    declare_parameter<int>("param3", 33333);
    declare_parameter<int>("param4", 0);
    declare_parameter<int>("param5", 0);
    declare_parameter<int>("max_loops", 3);

    log_parameter("param1");
    log_parameter("param2");
    log_parameter("param3");

    set_parameter(rclcpp::Parameter("param4", 4));
    set_parameter(rclcpp::Parameter("param5", 5));
    RCLCPP_INFO(get_logger(), "Param4 is set to be 4");
    RCLCPP_INFO(get_logger(), "Param5 is set to be 5");

    timer_ = create_wall_timer(1s, std::bind(&ParamDemoNode::tick, this));
  }

  bool done() const
  {
    return done_;
  }

private:
  void log_parameter(const std::string & name)
  {
    const int value = get_parameter(name).as_int();
    RCLCPP_INFO(get_logger(), "Get %s = %d", name.c_str(), value);
  }

  void tick()
  {
    ++loop_count_;
    const int max_loops = get_parameter("max_loops").as_int();

    if (loop_count_ == 1) {
      set_parameter(rclcpp::Parameter("param2", 2));
      RCLCPP_INFO(get_logger(), "Param2 updated to 2");
    }

    std::vector<std::string> names = {
      "max_loops", "param1", "param2", "param3", "param4"};
    names.push_back("param5");
    std::sort(names.begin(), names.end());

    std::string joined;
    for (size_t i = 0; i < names.size(); ++i) {
      joined += names[i];
      if (i + 1 < names.size()) {
        joined += ", ";
      }
    }

    RCLCPP_INFO(get_logger(), "=============Loop==============");
    RCLCPP_INFO(get_logger(), "param list: [%s]", joined.c_str());

    for (const auto & name : names) {
      const auto value = get_parameter(name);
      if (value.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER) {
        RCLCPP_INFO(
          get_logger(),
          "parameter %s = %lld",
          name.c_str(),
          static_cast<long long>(value.as_int()));
      }
    }

    if (loop_count_ >= max_loops) {
      RCLCPP_INFO(get_logger(), "Reached max_loops=%d, shutting down", max_loops);
      timer_->cancel();
      done_ = true;
    }
  }

  rclcpp::TimerBase::SharedPtr timer_;
  bool done_;
  int loop_count_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<ParamDemoNode>();
  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);
  while (rclcpp::ok() && !node->done()) {
    executor.spin_some();
    std::this_thread::sleep_for(100ms);
  }
  executor.remove_node(node);
  node.reset();
  rclcpp::shutdown();
  return 0;
}

File: ros2_ws/src/param_demo_py/param_demo_py/param_demo.py
import rclpy
from rclpy.node import Node


class ParamDemoNode(Node):
    def __init__(self) -> None:
        super().__init__("param_demo_py")
        self.loop_count = 0
        self.declare_parameter("param1", 111)
        self.declare_parameter("param2", 222)
        self.declare_parameter("param3", 33333)
        self.declare_parameter("param4", 0)
        self.declare_parameter("param5", 0)
        self.declare_parameter("max_loops", 3)

        self.get_logger().info(f"Get param1 = {self.get_parameter('param1').value}")
        self.get_logger().info(f"Get param2 = {self.get_parameter('param2').value}")
        self.get_logger().info(f"Get param3 = {self.get_parameter('param3').value}")

        self.set_parameters(
            [
                rclpy.parameter.Parameter("param4", value=4),
                rclpy.parameter.Parameter("param5", value=5),
            ]
        )
        self.timer = self.create_timer(1.0, self.tick)

    def tick(self) -> None:
        self.loop_count += 1
        max_loops = int(self.get_parameter("max_loops").value)
        if self.loop_count == 1:
            self.set_parameters([rclpy.parameter.Parameter("param2", value=2)])
            self.get_logger().info("Param2 updated to 2")

        if self.loop_count == 2 and self.has_parameter("param5"):
            self.undeclare_parameter("param5")
            self.get_logger().info("Param5 deleted via undeclare_parameter")

        names = sorted(self._parameters.keys())
        self.get_logger().info("=============Loop==============")
        self.get_logger().info(f"param list: {names}")
        for name in names:
            self.get_logger().info(f"parameter {name} = {self.get_parameter(name).value}")
        if self.loop_count >= max_loops:
            self.get_logger().info(f"Reached max_loops={max_loops}, shutting down")
            self.destroy_timer(self.timer)
            self.context.shutdown()


def main() -> None:
    rclpy.init()
    node = ParamDemoNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()
